Add mp-calendar component

mduan opened the pull request:
![image](https://cloud.githubusercontent.com/assets/198538/24164934/552e6260-0e2c-11e7-9966-cc5aae6ebe98.png)

mduan commented on lib/components/calendar/index.js:
```diff
@@ -0,0 +1,272 @@
+  }
+}
+
+document.registerElement(`mp-calendar`, Calendar);
```
Use `registerMPElement`

mduan replied:
Use `registerMPElement`

evnp commented on lib/components/calendar/index.js:
```diff
@@ -0,0 +1,270 @@
+  }
+
+  get maxDate() {
+    return this.getDateAttribute(`max-date`) || new Date();
```
Would it be possible to set no max date with this setup? Looks like explicitly setting `'max-date': null` would default to `new Date()`. Think I'd go with
```
return this.hasAttribute(`max-date`) ? this.getDateAttribute(`max-date`) : new Date();
```

evnp commented on lib/components/calendar/index.js:
```diff
@@ -0,0 +1,270 @@
+      return parseDate(this.getAttribute(name));
+    } else {
+      return null;
+    }
```
This is great, reduces a lot of code duplication below. `parseDate` will return `null` if the passed date is invalid, think it'd be worth throwing an error in that case? Seems a bit more developer-friendly since that's likely a mistake (you'd have to explicitly check for `'date-attr': null` since I'm guessing we still want that to e valid).

evnp commented on lib/components/calendar/index.js:
```diff
@@ -0,0 +1,270 @@
+
+  setDateAttribute(name, date) {
+    if (date) {
+      this.setAttribute(name, formatDate(date, {iso: true}));
```
Same here, think I'd make things explicit and throw an error if `formatDate` ever returns `null`

evnp commented on lib/components/calendar/index.js:
```diff
@@ -0,0 +1,270 @@
+      this.picker.setEndRange(this.selectedToDate);
+    }
+
+    if (this.selectedFromDate && !this.selectedToDate) {
```
This isn't you, but I'd actually make this check
```
!(this.selectedFromDate && this.selectedToDate) && (this.selectedFromDate || this.selectedToDate)
```
(or `!this.selectedFromDate ^ !this.selectedToDate`) to make this less dependent on the specific behavior of the widget - from always selected first.

mduan replied:
Neat. Never used xor operator in code before.

evnp commented on lib/util/date.js:
```diff
@@ -180,6 +180,31 @@ export function relativeToAbsoluteDate(relativeDateInt, unit, {utcOffset=null}={
 /**
+ * Accepts a [from, to] date range and the focused date ('to' or 'from') being edited
+ * @param {Array.<string>} - date range [from, to]
+ * @param {string} - one of <`from`, `to`>
```
Can we include param names here?
```
+ * @param {Array.<string>} dates - date range [from, to]
+ * @param {string} focusedDate - one of <`from`, `to`>
```

evnp replied:
outdated, disregard

evnp commented on lib/util/date.js:
```diff
@@ -180,6 +180,31 @@ export function relativeToAbsoluteDate(relativeDateInt, unit, {utcOffset=null}={
 
 /**
+ * Accepts a [from, to] date range and the focused date ('to' or 'from') being edited
+ * @param {Array.<string>} - date range [from, to]
```
Since this always needs an array of 2 as input, how about just making it 
```
normalizeDateRange(from, to, focused)
```
so the structure of the required input is simpler?

evnp commented on lib/util/date.js:
```diff
@@ -180,6 +180,31 @@ export function relativeToAbsoluteDate(relativeDateInt, unit, {utcOffset=null}={
 }
 
 /**
+ * Accepts a [from, to] date range and the focused date ('to' or 'from') being edited
```
This doesn't tell me what the function does, as a reader I'd like to see the intention summarized here

evnp commented on lib/util/date.js:
```diff
@@ -180,6 +180,31 @@ export function relativeToAbsoluteDate(relativeDateInt, unit, {utcOffset=null}={
+    }
+  }
+
+  return [from, to];
```
Is the desired behavior here to set both from and to to the same date if to comes before from or vice versa? That seems inferior to the current Insights behavior of swapping from<->to if they're in the wrong order (Jeff and I worked on this, was intentional). The swap seems more useful to the user, I don't think I'd ever want to select a range and then see them both move to the same date

mduan replied:
Discussed with Jeff. This is the desired behavior.

evnp commented on package.json:
```diff
@@ -76,6 +76,7 @@
     "mixpanel-browser": "2.9.16",
     "moment": "^2.17.1",
     "panel": "0.10.0",
+    "pikaday": "git://github.com/dbushell/Pikaday.git#a1704adef06e985a448cd8ab313dee5a5787a228",
```
ooh. What are we getting out of this revision?

mduan replied:
It looks to be so that we get this fix: https://github.com/dbushell/Pikaday/commit/a1704adef06e985a448cd8ab313dee5a5787a228. 

The latest npm release 1.5.1 (https://github.com/dbushell/Pikaday/releases/tag/1.5.1) which is before that commit.

evnp commented on lib/components/calendar/index.styl:
```diff
@@ -0,0 +1,182 @@
+
+          &:last-child
+            padding-right 18px
+
```
Thanks for cleaning this up, looks a lot better

evnp commented on lib/util/date.js:
```diff
@@ -180,6 +180,31 @@ export function relativeToAbsoluteDate(relativeDateInt, unit, {utcOffset=null}={
 
 /**
+ * Accepts a [from, to] date range and the focused date ('to' or 'from') being edited
+ * @param {Array.<string>} dates - date range [from, to]
```
Since this always needs an array of 2 as input, how about just making it 
```
normalizeDateRange(from, to, focused)
```
so the structure of the required input is simpler?

mduan commented:
@evnp Can you take another look. I've addressed your comments.

Within `mp-calendar`, I've also changed the code to use `string` dates (e.g. `2017-01-15`) rather than `Date` dates since the `string` format is more useful for what `mp-calendar` needs to do.

I've also added a bunch of wct tests.

evnp commented on examples/style-guide/index.js:
```diff
@@ -69,6 +69,27 @@ document.registerElement(`style-guide`, class extends Component {
+
+        getCalendarEventHandlers: () => {
+          return {
+            selectDate: e => console.log(e),
```
Debug code?

mduan commented on test/components/calendar.js:
```diff
@@ -0,0 +1,264 @@
+      this.resetCalendar({'min-date': `2017-01-05`}, () => {
+        this.navigateCalendar({year: 2017, month: 1});
+        const disabledDayEls = this.calendar.el.querySelectorAll(`td.is-disabled:not(.has-event) .pika-day`);
+        this.verifyDayEl(disabledDayEls[disabledDayEls.length - 1], {year: 2017, month: 1, day: 4});
```
Good idea. Done.
